src/xproc3/harness-lib.xpl

```diff
@@ -0,0 +1,293 @@
+       x:report, the error x:ERR001 is thrown.
+   -->
+   <p:declare-step type="x:format-report" name="format">
+      <!-- the port declarations -->
+      <p:input port="source" primary="true"/>
+      <p:output port="result" primary="true"/>
+
+      <p:option name="parameters" as="map(xs:QName,item()*)?"/>
+
+      <p:group>
+        <!-- param: xspec-home: the dir with the sources of XSpec if EXPath packaging
+             is not supported -->
+         <p:variable name="xspec-home" select="map:get($parameters, xs:QName('xspec-home'))"/>
+         
+         <p:variable name="formatter"
+            select="if ( $xspec-home ) then
+                  resolve-uri('src/reporter/format-xspec-report.xsl', $xspec-home)
+               else
+                  'http://www.jenitennison.com/xslt/xspec/format-xspec-report.xsl'"/>
+
+         <!-- log the report? -->
+         <x:log if-set="log-xml-report">
+            <p:with-option name="parameters" select="$parameters"/>         
+         </x:log>
+
+         <!-- if there is a report, format it, or it is an error -->
+         <p:choose>
+            <p:when test="exists(/x:report)">
+               <x:indent name="indent">
+                  <p:with-option name="parameters" select="$parameters"/>
+               </x:indent>
+               <p:load name="formatter" pkg:kind="xslt">
+                  <p:with-option name="href" select="$formatter"/>
+               </p:load>
+
+               <p:xslt name="format-report">
+                  <p:with-input port="source" pipe="@indent"/>
+                  <p:with-input port="stylesheet" pipe="@formatter"/>
+                  <p:with-option name="parameters" select="$parameters"/>
+               </p:xslt>
+            </p:when>
+
+            <p:otherwise>
+               <p:error code="x:ERR001">
+                  <p:with-input port="source">
+                     <p:inline>
+                        <message>Not a x:report document</message>
+                     </p:inline>
+                  </p:with-input>
+               </p:error>
+            </p:otherwise>
+         </p:choose>
+      </p:group>
+
+      <!-- log the report? -->
+      <x:log if-set="log-report">
+         <p:with-option name="parameters" select="$parameters"/>         
+      </x:log>
+   </p:declare-step>
+
+   <!-- Escapes markup. Also mimics @use-character-maps="x:disable-escaping" in
+      ../compiler/xquery/main.xsl. -->
+   <p:declare-step type="x:escape-markup" name="escape-markup">
+      <p:input  port="source" primary="true"/>
+      <p:output port="result" primary="true"/>
+
+      <p:cast-content-type content-type="text/plain"/>
+      <p:text-replace pattern="&#xE801;" replacement="&lt;"/>
+      <p:text-replace pattern="&#xE803;" replacement="&gt;"/>
+   </p:declare-step>
+
+   <!-- Serializes the source document with indentation and reloads it -->
+   <p:declare-step type="x:indent" name="indent">
```

Comment:
Is the purpose of the `x:indent` step just for logging the XML report in a human-readable format? I'm not suggesting changes but merely want to understand this code. If there's more to it than logging, can you explain?